R/combine_checkboxes.R

```diff
@@ -0,0 +1,192 @@
+
+  # Get metadata reference table, check that chosen fields are checkboxes
+  metadata <- get_metadata_ref(data_tbl_mod, supertbl, tbl, instrument_identifiers)
+
+  # Replace TRUEs/1s with raw/label values from metadata
+  data_tbl_mod <- data_tbl_mod %>%
+    mutate(across(
+      -c(any_of(instrument_identifiers), !!values_to),
+      ~ replace_true(.x,
+        cur_column(),
+        metadata = metadata,
+        raw_or_label = raw_or_label
+      )
+    ))
+
+  # Convert values_to from TRUE/FALSE to multi_value_label or identified single val
+  data_tbl_mod <- data_tbl_mod %>%
+    mutate(
+      across(field_names, as.character), # enforce to character strings
+      across(!!values_to, ~ as.character(.))
+    ) %>%
+    rowwise() %>%
+    mutate(
+      !!values_to := ifelse(!!sym(values_to) == "TRUE",
+        multi_value_label,
+        coalesce(!!!syms(field_names))
+      ),
+      !!values_to := ifelse(is.na(!!sym(values_to)), values_fill, !!sym(values_to))
+    ) %>%
+    ungroup() %>%
+    select(any_of(instrument_identifiers), !!values_to) %>%
+    mutate(
+      !!values_to := factor(!!sym(values_to), levels = c(metadata[[raw_or_label]], multi_value_label, values_fill))
+    )
+
+  # Join back onto original data tbl
+  data_tbl_mod <- data_tbl_mod %>%
+    right_join(data_tbl, by = intersect(instrument_identifiers, names(data_tbl_mod))) %>%
+    relocate(!!values_to, .after = everything())
+
+  # Keep or remove original multi columns
+  if (!keep) {
+    data_tbl_mod <- data_tbl_mod %>%
+      select(-field_names)
+  }
+
+  # Update the supertbl data tibble
+  supertbl$redcap_data[supertbl$redcap_form_name == tbl][[1]] <- data_tbl_mod
+
+  supertbl
+}
+
+#' @title Utility function for getting metadata raw and label values for checkboxes
+#'
+#' @param data a data tibble
+#' @param supertbl A supertibble generated by [read_redcap()].
+#' @param tbl The name of the REDCap form (instrument) containing the checkbox
+#' fields.
+#' @param instrument_identifiers Character string vector of project record identifier vars
+#'
+#' @returns a tibble
+#'
+#' @keywords internal
+get_metadata_ref <- function(data,
+                             supertbl,
+                             tbl,
+                             instrument_identifiers) {
+  # Create a metadata reference table linking field name to raw and label values
+  out <- supertbl$redcap_metadata[supertbl$redcap_form_name == tbl][[1]] %>%
+    filter(.data$field_name %in% names(data)[!names(data) %in% instrument_identifiers])
+
+  # Make sure selection is checkbox metadata field type
+  check_fields_are_checkboxes(out)
+
+  out <- out %>%
+    select(.data$field_name, .data$select_choices_or_calculations) %>%
+    mutate(
+      original_field = sub("___.*$", "", .data$field_name)
+    ) %>%
+    mutate(
+      pairs = strsplit(.data$select_choices_or_calculations, " \\| "),
+      label_value = NA,
+      label_value = purrr::map2_chr(pairs, row_number(), \(.x, .y) .x[.y])
+    )
```

Comment:
I don't love that we're reparsing `select_choices_or_calculations` here. Ideally we could use `parse_labels()` if not save the parsed values in the metadata inside `read_redcap()`.

Reply:
Hm, true it does seem a little duplicative but `parse_labels()` works slightly differently. It would be a matter of a dplyr chain like here versus some iterative operation to grab the `parse_labels()` dataframe output and then join back on to the metadata.

If we decided to update the metadata that what comes out of `read_redcap` includes something like `checkbox_raw` / `checkbox_label` alongside `select_choices_or_calculations` it would definitely make this easier to work with, but it also would be nonstandard with what `redcap_metadata_read()` returns.